Add support for configurable notification comments

mtreinish opened the pull request:
This commit adds a new feature for enabling per repo configurable
notification comments to qiskit-bot. When a repository has a per repo
configuration file (ie qiskit_bot.yaml in the repo root) with the
notifications section set this can be used to provide a list of
usernames that will get mentioned in a generated comment automatically
based on the files modified in the PR. This lets each project have a
local configuration of who gets notified of changes to each submodule in
the repository.

Fixes #5

mtreinish committed:
Add support for configurable notification comments

This commit adds a new feature for enabling per repo configurable
notification comments to qiskit-bot. When a repository has a per repo
configuration file (ie qiskit_bot.yaml in the repo root) with the
notifications section set this can be used to provide a list of
usernames that will get mentioned in a generated comment automatically
based on the files modified in the PR. This lets each project have a
local configuration of who gets notified of changes to each submodule in
the repository.

Fixes #5

mtreinish committed:
Fix Python 3.7 unittest compatibility

mtreinish committed:
Fix attribute access on PyGithub File objects

The filename attribute on PyGithub's File class is filename, not
file_name. This typo caused things to crash when trying to get the list
of changed files. This commit fixes this oversight.

mtreinish committed:
Finish default prelude text

mtreinish committed:
Fix lint and default prelude text

mtreinish commented:
Two things I just realized I should add to this before we merge. First I should add a section to the readme on the per repo `qiskit_bot.yaml` config file. Right now it's not clear how this can get used by projects to opt-in to notifications (or to configure the changelog labels which was a pre-existing feature). The other thing is I should add an option to have the bot always leave a comment, if there isn't a configured user for notification it can just print the prelude. This is useful for having initial communication to the user.

mtreinish committed:
Add option for always notifying even without matching files

This commit adds a new option to the local per repo config file for
always notifying on PR opening even if there are no matching paths
to mention users in the message. If this flag is set true it will always
at least print the configured prelude.

mtreinish committed:
Add README section on in-repo configuration

This commit adds a new section to the README documenting how to
configure qiskit-bot's behavior on a per repo level by creating a
`qiskit_bot.yaml` configuration file. It shows each option and how they
influence the behavior of the bot.

mtreinish commented:
For an example of this in practice see: https://github.com/mtreinish/qiskit-terra-sandbox/pull/15#issuecomment-1031945765 where I tested this with the bot in a local deployment.

manoelmarques commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+        if notify_list or always_notify:
+            default_prelude = """Thank you for opening a new pull request.
+
+Before your PR can be merged it will first need to run and pass continuous
```
Before your PR can be merged it will first need to run, pass continuous

mtreinish replied:
I'm not sure I understand the comma here. What about

```suggestion
Before your PR can be merged it will first need to pass continuous
```

the "run" here was redundant anyway.

jakelishman replied:
Your new form looks better to me. I think before it read like it was a list of three things, but the Oxford comma might have been more correct in that situation.

At any rate, the word "also" in "also be reviewed" is probably unnecessary.

manoelmarques commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+            default_prelude = """Thank you for opening a new pull request.
+
+Before your PR can be merged it will first need to run and pass continuous
+integration tests and be also be reviewed. Sometimes the review process can
```
integration tests and also be reviewed. Sometimes the review process can

mtreinish replied:
```suggestion
integration tests and also be reviewed. Sometimes the review process can
```

mtreinish committed:
Update qiskit_bot/notifications.py

Co-authored-by: Manoel Marques <manoelmrqs@gmail.com>

mtreinish committed:
Fix unittests after updating default notification prelude

In the previous commit we updated the default notification prelude to
fix a typo, but that commit neglected to update the tests accordingly.
This caused all the tests to error because the expected string didn't
match what was actually used. This commit fixes that oversight and fixes
the test failures.

jakelishman commented on README.md:
```diff
@@ -22,6 +22,13 @@ push a git tag to github. When coupled with pypi artifact (wheel and sdist)
 CI jobs pushing a tag becomes the only required manual step to push a release
 everything else will now be done automatically.
 
+The othe main feature the bot offers is to automatically leave a message on
```
```suggestion
The other main feature the bot offers is to automatically leave a message on
```

jakelishman commented on README.md:
```diff
@@ -44,3 +51,93 @@ the webhook to send all necessary event types to the endpoint where the bot
+contents. Currently this configuration file is used to control two things,
+the changelog generation behavior and whether the bot will leave comments
+on new pull requests when they're opened (and the exact behavior of that
+comment). n example of a fully populated configuration file is:
```
```suggestion
contents. Currently this configuration file is used to control two things:
the changelog generation behavior, and whether the bot will leave comments
on new pull requests when they're opened (and the exact behavior of that
comment). An example of a fully populated configuration file is:
```

jakelishman commented on README.md:
```diff
@@ -44,3 +51,93 @@ the webhook to send all necessary event types to the endpoint where the bot
+  PRs. While only a subset of people are authorized to approve pull requests for
+  merging everyone is encouraged to review open pull requests. Doing reviews
+  helps reduce the burden on the core team and helps make the project's code
+  better for everyone.
```
This might want updating with other typo fixes.

mtreinish replied:
Yeah, I'll update it here and in the tests after the last commit

jakelishman replied:
Just a reminder to update this.

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019
```
How much do we care about this stuff being accurate?  I think Nature have a checker for the copyrights, which might be nice to use for consistency across Qiskit in the future.

mtreinish replied:
It honestly isn't really enforceable for anything. Also `(C)` is meaningless it's not actually a copyright symbol.  It's just busy work imo to keep updating things on every when we have a complete git log to show modification history. For source distribution without the git history outside of the copyright statement itself the year doesn't really matter too much. That being said I can just fix this easily enough

```suggestion
# (C) Copyright IBM 2022
```

jakelishman replied:
Yeah, on qutip I literally just removed all these notices for that reason (qutip/qutip#1579).  The licence in the repo root is sufficient if it's distributed with the package.

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def trigger_notifications(pr_number, repo, conf):
+    """Do the post tag release processes."""
```
Is this docstring accurate?

mtreinish replied:
Nope, it was sloppy copy and paste. It should be something like:
```suggestion
def trigger_notifications(pr_number, repo, conf):
    """Process any potential notifications on a new PR."""
```

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+PRs. While only a subset of people are authorized to approve pull requests for
+merging everyone is encouraged to review open pull requests. Doing reviews
+helps reduce the burden on the core team and helps make the project's code
+better for everyone.
```
```suggestion
While you're waiting, please feel free to review other open
PRs. While only a subset of people are authorized to approve pull requests for
merging, everyone is encouraged to review open pull requests. Doing reviews
helps reduce the burden on the core team and helps make the project's code
better for everyone.
```

jakelishman replied:
(I doubt many people are going to read it, but probably best be as concise as possible to improve our odds.)

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+merging everyone is encouraged to review open pull requests. Doing reviews
+helps reduce the burden on the core team and helps make the project's code
+better for everyone.
+"""
```
It's probably going to be nicer for users if we reflow this into continuous text, rather than dumping out a hard-wrapped block.

jakelishman commented on README.md:
````diff
@@ -44,3 +51,93 @@ the webhook to send all necessary event types to the endpoint where the bot
+  ```
+
+- `notifications`: This contains a mapping of path regexes to a list of usernames
+  to notify if an opened PR touches files that match a particular regex (as found
+  by Python's stdlib `re.search()` function). For example if you set the path regex
+  to `".*"` this would match everything, but using a regex gives you control over
````
It might be better to use/mention single quotes for regex strings - my understanding is that single-quoted YAML strings don't parse escape characters, but double-quotes do, and literal `\` is needed in regexes fairly commonly.

mtreinish replied:
Yeah, that's a fair point I had forgot that detail from the yaml spec (https://yaml.org/spec/1.2.2/#double-quoted-style vs https://yaml.org/spec/1.2.2/#single-quoted-style for reference). I'll update this to mention that

jakelishman replied:
Don't know if you want to bother with this (explaining single quotes for YAML regexes).  On a second reading, I'm not sure this README is the right place to include a YAML tutorial.

jakelishman reviewed (commented):
Are we planning to use this sort of user notification on Terra/other parts of Qiskit?  What's the advantage of this user notification over the current `CODEOWNERS`?

mtreinish commented:
> Are we planning to use this sort of user notification on Terra/other parts of Qiskit? What's the advantage of this user notification over the current `CODEOWNERS`?

Yeah, that's the basic idea. The biggest advantage is that it decouples notification for a review request from merge permission. A codeowner entry always gets you both, but there might be people that want to explicitly get pinged for PRs for some subsection of the repo but don't have write permission yet. Another category is for splitting up notification of a core team, if the core team has merge permission on the repo but would only like to actively be pinged as an individual on a subtree this can be leveraged for that. To a smaller degree the advantage this provides is the regex match is more flexible than the selection the codeowners file does but I'm not sure how many people are going to want to leverage that (personally I'm looking forward to being notified on every file that has two `'a'`s in the path :P)

jakelishman commented:
Yeah, that sounds like solid reasoning to me.  I guess it doesn't affect me (or you) personally too much, since we're already subscribed to all PRs made, but I think it'll be a lot of help for other people in the team.

mtreinish commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,81 @@
+            default_prelude = """Thank you for opening a new pull request.
+
+Before your PR can be merged it will first need to run and pass continuous
+integration tests and also be reviewed. Sometimes the review process can
```
```suggestion
integration tests and be reviewed. Sometimes the review process can
```

mtreinish committed:
Fix typos found in code review

This commit fixes typos in various documentation, text blocks,
and docstrings which were found during code review.

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Update copyright year in qiskit_bot/notifications.py

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Update tests for recent prelude changes

mtreinish committed:
Merge branch 'master' into add-notifications

mtreinish committed:
Remove fixed 80 character line wrap from default notification message

jakelishman commented on README.md:
```diff
@@ -44,3 +51,93 @@ the webhook to send all necessary event types to the endpoint where the bot
+qiskit-bot gives projects some local configuration options that can be set in
+the repository. To set a local configuration file a file `qiskit_bot.yaml` must
+be created in the root of the git repository. If this file is present then
+qiskit-bot will read it before every action and adjust behavior based on it's
```
```suggestion
qiskit-bot will read it before every action and adjust behavior based on its
```

jakelishman commented on README.md:
```diff
@@ -44,3 +51,93 @@ the webhook to send all necessary event types to the endpoint where the bot
+  to notify if an opened PR touches files that match a particular regex (as found
+  by Python's stdlib `re.search()` function). For example if you set the path regex
+  to `".*"` this would match everything, but using a regex gives you control over
+  exactly how and what matches a particular group. If a path matches the listed
+  usernames will be listed in the notification comment left by the bot on a new PR
+  being opened. This is an additive, so if there is more than 1 match the users
```
Did you also consider having the bot assign the issue instead of a notification comment?  I guess assignment feels more natural, but can't apply to teams.

mtreinish replied:
I did, but the main issue with this is that assignment is limited to project members only and if anyone wanted to receive a notification outside of being an explicit project team member assignment was not really an option

jakelishman replied:
Yep, sounds good.  I'd assumed you'd thought about it, just wondering the reasons.  Now you mention it, I think I knew this reason already sorry.

jakelishman commented on qiskit_bot/config.py:
```diff
@@ -62,21 +62,26 @@ def load_config(path):
 local_config_schema = vol.Schema({
-    'users': [{}],
-    'categories': vol.Optional({}),
+    vol.Optional('categories', default=default_changelog_categories): dict,
+    vol.Optional('notifications'): {vol.Extra: [str]},
+    vol.Optional('notification_prelude'): str,
+    vol.Optional('always_notify'): bool,
```
This doesn't matter, just a question: are there any reasons to define `categories` with a default here, but leave the others default-less?

mtreinish replied:
If there is a sane implicit default value not really, just sloppy style on my part mostly. I think I might have had some trouble with voluptuous populating defaults at some point and just opted not to use it if I couldn't get it to work reliably (but that might have just been user error, I don't really remember)

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,86 @@
+
+LOG = logging.getLogger(__name__)
+
+with io.StringIO() as buf:
+    buf.write("Thank you for opening a new pull request.\n\n")
+    buf.write(
+        "Before your PR can be merged it will first need to pass "
+        "continuous integration tests and be reviewed. Sometimes "
+        "the review process can be slow, so please be patient.\n\n"
+    )
+    buf.write(
+        "While you're waiting, please feel free to review other "
+        "open PRs. While only a subset of people are authorized "
+        "to approve pull requests for merging, everyone is "
+        "encouraged to review open pull requests. Doing reviews "
+        "helps reduce the burden on the core team and helps make "
+        "the project's code better for everyone.\n"
+    )
+    DEFAULT_PRELUDE = buf.getvalue()
```
Again, not important at all: is there an advantage to defining this here rather than as part of the schema/configuration file?

(I'm guessing testing?)

mtreinish replied:
Yeah, it's mostly testing I think, but I honestly wrote this long enough ago I'm not really sure

jakelishman commented on qiskit_bot/notifications.py:
```diff
@@ -0,0 +1,86 @@
+        file_list = pr.get_files()
+        filenames = [file.filename for file in file_list]
+        for path_regex, user_list in notification_regex.items():
+            for file_name in filenames:
+                if path_regex.search(file_name):
```
I think this probably is irrelevant, because we'll always be running the bot on Linux machines, but is there a potential path-separator issue brewing here on Windows?  If you want to match a subdirectory, I don't know what kind of path names the GitHub interface is going to return - will they be `/`-separated or `\`-separated on Windows?

(I'm also happy just to ignore the comment and say "the bot only runs on unix-likes".)

mtreinish replied:
At least in production we only run this on linux (specifically debain bullseye using the official python docker image as the base). I'm not sure there is value in trying to support this on windows as I find it extremely unlikely that anyone would want to deploy the bot to a windows environment.

That being said I did think about this a long time ago and made the decision to only list linux as a supported environment in the package metadata trove classifiers: https://github.com/Qiskit/qiskit-bot/blob/master/setup.cfg#L13 although I don't know how many people actually look at those

jakelishman reviewed (commented):
This all looks fine to me.  Aside from the typo fix, most of the comments are minor questions rather than fixes.

mtreinish committed:
Update README.md

Co-authored-by: Jake Lishman <jake@binhbar.com>

mtreinish committed:
Update default message in README
